File: bug-intake-agent.py
#!/usr/bin/env python3
"""
Bug Intake Agent - Syncs Intercom bug reports to Linear issue tracker.

This script:
1. Searches Intercom for conversations tagged "Bug Report" updated in the last 60 minutes
2. Filters out conversations already synced (have "linear-synced" tag)
3. Posts a comment to Linear ISSUE-281 with the bug details
4. Marks conversations as synced to prevent reprocessing
"""

import os
import sys
import json
import requests
from datetime import datetime, timedelta
from typing import Optional

class IntercomClient:
    """Intercom API client"""
    
    def __init__(self, api_token: str):
        self.base_url = "https://api.intercom.io"
        self.headers = {
            "Authorization": f"Bearer {api_token}",
            "Accept": "application/json",
            "Content-Type": "application/json"
        }
    
    def search_conversations(self, query: dict) -> dict:
        """Search for conversations"""
        url = f"{self.base_url}/conversations/search"
        response = requests.post(url, json=query, headers=self.headers)
        response.raise_for_status()
        return response.json()
    
    def get_conversation(self, conversation_id: str) -> dict:
        """Get full conversation details"""
        url = f"{self.base_url}/conversations/{conversation_id}"
        response = requests.get(url, headers=self.headers)
        response.raise_for_status()
        return response.json()
    
    def get_contact(self, contact_id: str) -> dict:
        """Get contact details"""
        url = f"{self.base_url}/contacts/{contact_id}"
        response = requests.get(url, headers=self.headers)
        response.raise_for_status()
        return response.json()
    
    def add_tag_to_conversation(self, conversation_id: str, tag_name: str) -> dict:
        """Add a tag to a conversation"""
        url = f"{self.base_url}/conversations/{conversation_id}/tags"
        data = {"tag": {"name": tag_name}}
        response = requests.post(url, json=data, headers=self.headers)
        response.raise_for_status()
        return response.json()
    
    def add_note_to_conversation(self, conversation_id: str, note_body: str, is_private: bool = True) -> dict:
        """Add an internal note to a conversation"""
        url = f"{self.base_url}/conversations/{conversation_id}/notes"
        data = {
            "body": note_body,
            "admin_id": None  # System note
        }
        response = requests.post(url, json=data, headers=self.headers)
        response.raise_for_status()
        return response.json()


class LinearClient:
    """Linear API client"""
    
    def __init__(self, api_token: str):
        self.base_url = "https://api.linear.app/graphql"
        self.headers = {
            "Authorization": f"Bearer {api_token}",
            "Content-Type": "application/json"
        }
    
    def create_comment(self, issue_id: str, body: str) -> dict:
        """Create a comment on an issue"""
        query = """
        mutation CreateComment($input: CommentCreateInput!) {
            commentCreate(input: $input) {
                comment {
                    id
                    url
                }
            }
        }
        """
        variables = {
            "input": {
                "issueId": issue_id,
                "body": body
            }
        }
        response = requests.post(
            self.base_url,
            json={"query": query, "variables": variables},
            headers=self.headers
        )
        response.raise_for_status()
        data = response.json()
        if "errors" in data:
            raise Exception(f"GraphQL error: {data['errors']}")
        return data["data"]["commentCreate"]["comment"]


def extract_bug_summary_and_details(conversation: dict) -> tuple:
    """
    Extract a one-line summary and repro steps from conversation.
    Returns (summary, repro_steps, attachment_urls)
    """
    parts = conversation.get("conversation_parts", {}).get("conversation_parts", [])
    
    # Get the initial issue from source
    source = conversation.get("source", {})
    initial_body = source.get("body", "").replace("<p>", "").replace("</p>", "").strip()
    
    # Find user responses and attachments
    user_id = source.get("author", {}).get("id")
    user_comments = []
    attachment_urls = []
    
    for part in parts:
        if part.get("author", {}).get("id") == user_id and part.get("part_type") == "comment":
            body = part.get("body", "").replace("<p>", "").replace("</p>", "").strip()
            if body:
                user_comments.append(body)
        
        # Collect attachment URLs
        for attachment in part.get("attachments", []):
            if attachment.get("url"):
                attachment_urls.append(attachment["url"])
    
    # Create summary - should be action-oriented and specific
    # First, try to extract from AI Title if available
    custom_attrs = conversation.get("custom_attributes", {})
    ai_title = custom_attrs.get("AI Title", "")
    
    if ai_title:
        summary = ai_title
    else:
        # Create summary from initial message
        if "loading" in initial_body.lower() and "buffer" in initial_body.lower():
            summary = "Monsha loading/buffering on workspace startup"
        elif "version" in initial_body.lower() and "test" in initial_body.lower():
            summary = "Cannot create additional versions of existing tests"
        else:
            summary = initial_body[:100].rstrip(".")
    
    # Get repro steps from user comments
    repro_steps = "\n".join(user_comments[:2]) if user_comments else ""
    
    return summary, repro_steps, attachment_urls


def sync_bug_to_linear(
    intercom_client: IntercomClient,
    linear_client: LinearClient,
    conversation: dict,
    contact: dict
) -> Optional[str]:
    """
    Sync a single bug conversation to Linear.
    Returns the Linear comment URL if successful, None otherwise.
    """
    # Check if already synced
    tags = conversation.get("tags", {}).get("tags", [])
    tag_names = [tag.get("name") for tag in tags]
    
    if "linear-synced" in tag_names:
        print(f"Skipping conversation {conversation['id']} - already synced")
        return None
    
    # Extract details
    summary, repro_steps, attachment_urls = extract_bug_summary_and_details(conversation)
    
    # Prepare comment for Linear
    user_name = contact.get("name", "Unknown")
    user_email = contact.get("email", "unknown@example.com")
    intercom_url = f"https://app.intercom.com/a/inbox/_/inbox/conversation/{conversation['id']}"
    
    # Build markdown comment
    comment_lines = [
        f"**{summary}** — [{user_name}]({intercom_url}) ({user_email})",
        ""
    ]
    
    if repro_steps:
        comment_lines.append(repro_steps)
        comment_lines.append("")
    
    if attachment_urls:
        comment_lines.append("Attachments:")
        for url in attachment_urls:
            comment_lines.append(f"- {url}")
        comment_lines.append("")
    
    comment_lines.append("@adel")
    
    comment_body = "\n".join(comment_lines)
    
    # Ensure comment stays under 150 words (approximate)
    # This is already handled by the summary structure
    
    try:
        # Post to Linear
        linear_comment = linear_client.create_comment("ISSUE-281", comment_body)
        linear_url = linear_comment["url"]
        
        print(f"✓ Posted comment to Linear: {linear_url}")
        
        # Apply linear-synced tag
        try:
            intercom_client.add_tag_to_conversation(conversation["id"], "linear-synced")
            print(f"✓ Tagged conversation {conversation['id']} with 'linear-synced'")
        except Exception as e:
            print(f"✗ Failed to tag conversation: {e}")
        
        # Add internal note
        try:
            note_body = f"Logged on Linear ISSUE-281. Link: {linear_url}"
            intercom_client.add_note_to_conversation(conversation["id"], note_body)
            print(f"✓ Added internal note to conversation {conversation['id']}")
        except Exception as e:
            print(f"✗ Failed to add note: {e}")
        
        return linear_url
        
    except Exception as e:
        print(f"✗ Failed to sync conversation {conversation['id']}: {e}")
        return None


def main():
    """Main entry point"""
    # Get API tokens from environment
    intercom_token = os.getenv("INTERCOM_API_TOKEN")
    linear_token = os.getenv("LINEAR_API_TOKEN")
    
    if not intercom_token or not linear_token:
        print("Error: INTERCOM_API_TOKEN and LINEAR_API_TOKEN environment variables required")
        sys.exit(1)
    
    # Initialize clients
    intercom = IntercomClient(intercom_token)
    linear = LinearClient(linear_token)
    
    # Calculate time window
    now = datetime.utcnow()
    one_hour_ago = now - timedelta(hours=1)
    
    print(f"Searching for bug reports updated in last 60 minutes (since {one_hour_ago.isoformat()}Z)")
    
    try:
        # Search for conversations with "Bug Report" tag
        search_query = {
            "query": {
                "operator": "AND",
                "value": [
                    {
                        "field": "updated_at",
                        "operator": ">=",
                        "value": int(one_hour_ago.timestamp())
                    },
                    {
                        "field": "tag",
                        "operator": "=",
                        "value": "Bug Report"
                    }
                ]
            }
        }
        
        results = intercom.search_conversations(search_query)
        conversations = results.get("conversations", [])
        
        print(f"Found {len(conversations)} bug-tagged conversations")
        
        synced_count = 0
        for conv in conversations:
            conv_id = conv.get("id")
            contact_id = conv.get("contact_ids", [])[0] if conv.get("contact_ids") else None
            
            if not contact_id:
                print(f"Skipping conversation {conv_id} - no contact found")
                continue
            
            # Get full details
            full_conv = intercom.get_conversation(conv_id)
            contact = intercom.get_contact(contact_id)
            
            # Sync to Linear
            result = sync_bug_to_linear(intercom, linear, full_conv, contact)
            if result:
                synced_count += 1
        
        if synced_count == 0:
            print("No new bug reports to sync")
        else:
            print(f"✓ Synced {synced_count} bug report(s)")
    
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: bug-intake-agent.js
#!/usr/bin/env node

/**
 * Bug Intake Agent - Hourly automation
 * Fetches Intercom bug reports and forwards them to Linear ISSUE-281
 */

const https = require('https');

// Configuration
const INTERCOM_TOKEN = process.env.INTERCOM_TOKEN;
const LINEAR_TOKEN = process.env.LINEAR_TOKEN;
const LINEAR_ISSUE_ID = 'ISSUE-281';
const BUG_REPORT_TAG = 'Bug Report';
const LINEAR_SYNCED_TAG = 'linear-synced';

if (!INTERCOM_TOKEN || !LINEAR_TOKEN) {
  console.error('Missing required environment variables: INTERCOM_TOKEN and LINEAR_TOKEN');
  process.exit(1);
}

/**
 * Make HTTPS request helper
 */
function makeRequest(options, body = null) {
  return new Promise((resolve, reject) => {
    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => {
        data += chunk;
      });
      res.on('end', () => {
        try {
          const parsed = data ? JSON.parse(data) : {};
          resolve({ status: res.statusCode, headers: res.headers, body: parsed });
        } catch (e) {
          resolve({ status: res.statusCode, headers: res.headers, body: data });
        }
      });
    });

    req.on('error', reject);
    if (body) {
      req.write(JSON.stringify(body));
    }
    req.end();
  });
}

/**
 * Search Intercom conversations with "Bug Report" tag updated in last 60 minutes
 */
async function fetchBugReportConversations() {
  const sixtyMinutesAgo = Math.floor((Date.now() - 60 * 60 * 1000) / 1000);

  const options = {
    hostname: 'api.intercom.io',
    path: '/conversations/search',
    method: 'POST',
    headers: {
      Authorization: `Bearer ${INTERCOM_TOKEN}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
  };

  // Use Intercom's conversation search with DSL
  const body = {
    query: {
      operator: 'AND',
      value: [
        {
          field: 'tag_names',
          operator: '=',
          value: BUG_REPORT_TAG,
        },
        {
          field: 'updated_at',
          operator: '>',
          value: sixtyMinutesAgo,
        },
      ],
    },
  };

  try {
    const response = await makeRequest(options, body);
    if (response.status !== 200) {
      console.error('Intercom search failed:', response.status, response.body);
      return [];
    }
    return response.body.conversations || [];
  } catch (error) {
    console.error('Error fetching bug report conversations:', error);
    return [];
  }
}

/**
 * Get full conversation details including all messages
 */
async function getConversationDetails(conversationId) {
  const options = {
    hostname: 'api.intercom.io',
    path: `/conversations/${conversationId}`,
    method: 'GET',
    headers: {
      Authorization: `Bearer ${INTERCOM_TOKEN}`,
      Accept: 'application/json',
    },
  };

  try {
    const response = await makeRequest(options);
    if (response.status !== 200) {
      console.error(`Failed to fetch conversation ${conversationId}:`, response.status);
      return null;
    }
    return response.body;
  } catch (error) {
    console.error(`Error fetching conversation details:`, error);
    return null;
  }
}

/**
 * Get contact details (user info)
 */
async function getContact(contactId) {
  const options = {
    hostname: 'api.intercom.io',
    path: `/contacts/${contactId}`,
    method: 'GET',
    headers: {
      Authorization: `Bearer ${INTERCOM_TOKEN}`,
      Accept: 'application/json',
    },
  };

  try {
    const response = await makeRequest(options);
    if (response.status !== 200) {
      console.error(`Failed to fetch contact ${contactId}:`, response.status);
      return null;
    }
    return response.body;
  } catch (error) {
    console.error(`Error fetching contact:`, error);
    return null;
  }
}

/**
 * Check if conversation already has "linear-synced" tag
 */
function isAlreadySynced(conversation) {
  const tags = conversation.tags?.data || [];
  return tags.some((tag) => tag.name === LINEAR_SYNCED_TAG);
}

/**
 * Extract thread content (initial message + replies)
 */
function extractThreadContent(conversation) {
  let content = {
    summary: null,
    repoSteps: null,
    attachmentUrls: [],
  };

  // Extract from conversation source (initial message)
  if (conversation.source && conversation.source.body) {
    content.summary = conversation.source.body.substring(0, 150);
  }

  // Extract from conversation parts (messages and attachments)
  const parts = conversation.conversation_parts?.conversation_parts || [];

  let textContent = [];
  let attachments = [];

  for (const part of parts) {
    if (part.part_type === 'comment' && part.body) {
      textContent.push(part.body);
    }

    // Extract attachment URLs
    if (part.attachments) {
      for (const attachment of part.attachments) {
        if (attachment.url) {
          attachments.push({
            name: attachment.name || 'Attachment',
            url: attachment.url,
          });
        }
      }
    }
  }

  // Look for "repro" or "steps" keywords in the text
  const fullText = textContent.join('\n');
  const repoMatch = fullText.match(/(?:repro|steps|reproduce)[\s:]+([^\n]+(?:\n[^\n]+)*)/i);
  if (repoMatch) {
    content.repoSteps = repoMatch[1].substring(0, 150);
  }

  content.attachmentUrls = attachments;
  return content;
}

/**
 * Post comment to Linear issue
 */
async function postLinearComment(issueId, commentBody) {
  const options = {
    hostname: 'api.linear.app',
    path: '/graphql',
    method: 'POST',
    headers: {
      Authorization: `Bearer ${LINEAR_TOKEN}`,
      'Content-Type': 'application/json',
    },
  };

  const query = `
    mutation CreateComment($issueId: String!, $body: String!) {
      commentCreate(input: {
        issueId: $issueId
        body: $body
      }) {
        comment {
          id
          url
        }
      }
    }
  `;

  const body = {
    query,
    variables: {
      issueId,
      body: commentBody,
    },
  };

  try {
    const response = await makeRequest(options, body);
    if (response.status !== 200 || response.body.errors) {
      console.error('Linear API error:', response.body.errors || response.body);
      return null;
    }
    return response.body.data?.commentCreate?.comment?.url || null;
  } catch (error) {
    console.error('Error posting to Linear:', error);
    return null;
  }
}

/**
 * Apply tag to Intercom conversation
 */
async function applyTag(conversationId, tagName) {
  const options = {
    hostname: 'api.intercom.io',
    path: `/conversations/${conversationId}/tags`,
    method: 'POST',
    headers: {
      Authorization: `Bearer ${INTERCOM_TOKEN}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
  };

  const body = {
    tag: {
      name: tagName,
    },
  };

  try {
    const response = await makeRequest(options, body);
    if (response.status !== 200) {
      console.error(`Failed to apply tag to conversation ${conversationId}:`, response.status);
      return false;
    }
    return true;
  } catch (error) {
    console.error(`Error applying tag:`, error);
    return false;
  }
}

/**
 * Add internal note to Intercom conversation
 */
async function addInternalNote(conversationId, noteBody) {
  const options = {
    hostname: 'api.intercom.io',
    path: `/conversations/${conversationId}/parts`,
    method: 'POST',
    headers: {
      Authorization: `Bearer ${INTERCOM_TOKEN}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
  };

  const body = {
    message_type: 'note',
    body: noteBody,
  };

  try {
    const response = await makeRequest(options, body);
    if (response.status !== 201) {
      console.error(`Failed to add note to conversation ${conversationId}:`, response.status);
      return false;
    }
    return true;
  } catch (error) {
    console.error(`Error adding internal note:`, error);
    return false;
  }
}

/**
 * Format comment for Linear
 */
function formatLinearComment(summary, contact, repoSteps, attachmentUrls, intercomUrl, email) {
  const lines = [];

  // Summary with user link
  const userName = contact.name || 'Unknown';
  lines.push(`**${summary}** — [${userName}](${intercomUrl}) (${email})`);
  lines.push('');

  // Repro steps or quoted text
  if (repoSteps) {
    lines.push(repoSteps);
    lines.push('');
  }

  // Attachment links
  if (attachmentUrls.length > 0) {
    for (const attachment of attachmentUrls) {
      lines.push(`[${attachment.name}](${attachment.url})`);
    }
    lines.push('');
  }

  // Mention
  lines.push('@adel');

  return lines.join('\n');
}

/**
 * Main agent function
 */
async function runAgent() {
  console.log('Starting bug intake agent...');

  // Step 1: Fetch bug report conversations from last 60 minutes
  console.log('Fetching bug report conversations from the last 60 minutes...');
  const conversations = await fetchBugReportConversations();

  if (conversations.length === 0) {
    console.log('No new bug report conversations found. Exiting silently.');
    return;
  }

  console.log(`Found ${conversations.length} bug report conversation(s)`);

  // Step 2-6: Process each conversation
  for (const conversation of conversations) {
    const conversationId = conversation.id;
    const conversationUrl = `https://app.intercom.com/a/inbox/${conversationId}`;

    console.log(`\nProcessing conversation ${conversationId}...`);

    // Check if already synced
    if (isAlreadySynced(conversation)) {
      console.log(`Conversation ${conversationId} already synced. Skipping.`);
      continue;
    }

    // Get full conversation details
    const fullConversation = await getConversationDetails(conversationId);
    if (!fullConversation) {
      console.error(`Failed to get details for conversation ${conversationId}`);
      continue;
    }

    // Get contact details (user info)
    const contactId = conversation.participants?.[0]?.id;
    if (!contactId) {
      console.error(`No contact found for conversation ${conversationId}`);
      continue;
    }

    const contact = await getContact(contactId);
    if (!contact) {
      console.error(`Failed to get contact details for ${contactId}`);
      continue;
    }

    // Extract thread content
    const threadContent = extractThreadContent(fullConversation);
    const summary = threadContent.summary || 'Bug report';
    const email = contact.email || 'unknown@email.com';
    const userName = contact.name || 'Unknown';
    const userId = contact.external_id || contact.id;

    console.log(`User: ${userName} (${userId}, ${email})`);
    console.log(`Summary: ${summary}`);

    // Format and post to Linear
    const commentBody = formatLinearComment(
      summary,
      contact,
      threadContent.repoSteps,
      threadContent.attachmentUrls,
      conversationUrl,
      email
    );

    console.log('Posting to Linear...');
    const linearCommentUrl = await postLinearComment(LINEAR_ISSUE_ID, commentBody);

    if (!linearCommentUrl) {
      console.error(`Failed to post comment to Linear for conversation ${conversationId}`);
      continue;
    }

    console.log(`Comment posted: ${linearCommentUrl}`);

    // Apply linear-synced tag
    console.log('Applying linear-synced tag...');
    const tagApplied = await applyTag(conversationId, LINEAR_SYNCED_TAG);
    if (!tagApplied) {
      console.error(`Failed to apply tag to conversation ${conversationId}`);
      continue;
    }

    // Add internal note
    const internalNote = `Logged on Linear ${LINEAR_ISSUE_ID}. Link: ${linearCommentUrl}`;
    console.log('Adding internal note...');
    const noteAdded = await addInternalNote(conversationId, internalNote);
    if (!noteAdded) {
      console.error(`Failed to add internal note to conversation ${conversationId}`);
      continue;
    }

    console.log(`Successfully processed conversation ${conversationId}`);
  }

  console.log('\nBug intake agent completed.');
}

// Run the agent
runAgent().catch((error) => {
  console.error('Fatal error in bug intake agent:', error);
  process.exit(1);
});
